### ISSUE TITLE
There is no need for putting data type while querying to dynamoDb

### ISSUE THREAD

**Issue by e-nouri** [ginger-io/beyonce#20]
I made a query using a partionkey, the constructed json query outputted by the beautiful Beyoncé is like this:

```
[AWS dynamodb 200 0.007s 0 retries] query({ TableName: 'test-dev',
  KeyConditionExpression: '#pk = :pk',
  ExpressionAttributeNames: { '#pk': 'pk' },
  ExpressionAttributeValues: { ':pk': { S: 'USR-test' } } })

```
Why not just put `ExpressionAttributeValues: { ':pk': 'USR-test'  }` without the type ?

**Comment by jcarver989** [comment]
Ah, so the types you see in `ExpressionAttributeValues` aren't added by Beyoncé directly (see here: https://github.com/ginger-io/beyonce/blob/d202b8de0dfbcbe1cf237892bcdf60575825bb06/src/main/dynamo/QueryBuilder.ts#L201)

These types actually get added by the `DynamoDB.DocumentClient` that the library uses internally (https://docs.aws.amazon.com/AWSJavaScriptSDK/latest/AWS/DynamoDB/DocumentClient.html).

**Comment by e-nouri** [comment]
Yes I just realised that after inspecting the requests.

For the taggedModels, if I am querying the table, why not add the "begins_with" for the prefix that I already provided, is there a reason for returning all the models even though I am using the AutherPartition  and not all the models partition ?

Also there is no beging_with() in Operator, do you accept PR for this ?

**Comment by jcarver989** [comment]
> ...there is no beging_with() in Operator, do you accept PR for this ?
Certainly - I just added an `ExpressionBuilder` in the last PR, so you'd want to add it here: 
https://github.com/ginger-io/beyonce/pull/21/files#diff-f73e911212402cc42f124f9e5f5ac3da

> For the taggedModels, if I am querying the table, why not add the "begins_with" for the prefix that I already provided, is there a reason for returning all the models even though I am using the AutherPartition and not all the models partition ?

I'm not quite sure I understand your question - could you clarify it? 

For context, the XXXPartition objects are designed to help you retrieve multiple types of models using a `.query` call. For example if you have an `Author` model and a `Book` model and stick them both into an `AuthorPartition` - then `beyonce.query(AuthorPartition.key({ ... })).exec()` would give you back both the author and the author's books (by design). 

If you wanted to retrieve just the author, you'd want to use `beyonce.get(AuthorModel.key({ ... })`. And if you wanted to filter out a model type from the results you'd do `beyonce.query(...).where("model", "<>", ModelType.Foo). On that last one, the types aren't smart enough (yet) to narrow the return type down (i.e. remove the `Foo: Foo[]` key from the results).

**Comment by e-nouri** [comment]
In my test right now, a user has his own profil and multiple cards and devices, so i use pk and sk to store them on the same partition but i prefix the sk with 'CRD', 'USR', 'DVS'.

When querying only the cards using the card model like so:
```
            let cards = await db
                .query(CardModel.partitionKey({ userId: user.id }))
                .exec()
```

I get back:
`{ Card:[...], Device:[...], User:[...] }`

Why not use the begin_with('CRD-') to filter out Device and User models and avoid the additional reads, since I am explicitly using the CardModel and not the Main Partition, does that make sense ?

That way if I want to query the 3 models I can use the MainPartition that sends back all the models, but if I want to avoid the additional reads, I am not talking about filtering, i can query each model efficiently.

DynamoDb filters out after reads, which means the read capacities you consumed are lost, it makes sense when the set that you are filtering out is small like 3 from 100 records, but it does not make much sense doing it the other way, filtering 100 records to get just 3. I hope I am making sense.

**Comment by jcarver989** [comment]
> Why not use the begin_with('CRD-') to filter out Device and User models and avoid the additional reads, since I am explicitly using the CardModel and not the Main Partition, does that make sense ?

Yup, makes total sense. You want to query for N instances of a single model type that live under a given partition key using a `KeyConditionExpression`. 

This turned out to be pretty easy to add -- so I think this should cover your use case: 
https://github.com/ginger-io/beyonce/pull/22

And it should be available for use in version `0.0.30`

**Comment by e-nouri** [comment]
Wonderful ! Thank you Joshua !

I am working on pagination right now, I saw the limit on the query but I am not sure if the lastEvaluatedKey is returned along with the models.  I saw that you use it to chunk read all the records, but imagine in my case instead of cards, I have orders, it is not 10 records, more like 100, so pagination is a must in this case.

Is there a way to make this behaviour configurable, if the limit given to the query is bigger than the internal limit, use the actual pagination strategy, otherwise do the old looping ?

**Comment by jcarver989** [comment]
> Is there a way to make this behaviour configurable...?

Certainly. 

I think the best experience for the caller is to abstract away the `lastEvaluatedKey` and instead offer an `iterator()` method that returns an async iterator with an opaque cursor.

I've added that in this PR https://github.com/ginger-io/beyonce/pull/23 - which is included in 0.0.31. 

So now you can do: 

```TypeScript
const iterator = beyonce
  .query(AuthorPartition.key({ id: "1" }))
  .iterator({ pageSize: 1 })

const { value: firstPage, done } = await iterator.next()
firstPage.items // GroupedModels<Author | Book>
firstPage.cursor // opaque cursor

if (!done) {
 const { value: secondPage} = iterator.next({ cursor: firstPage.cursor })
}
```

Or

```TypeScript
const iterator = beyonce
  .query(AuthorPartition.key({ id: "1" }))
  .iterator({ pageSize: 1 })

for await (const { items } of iterator) {
  ... 
}
```

**Comment by e-nouri** [comment]
Thank you Joshua, but Imagine I am getting only the second page (30 record) starting from LastEvalutedKey, this won't work.

This behaviour is good one reading data to do some batch, or some kinda of effecient processing without eating much memory, but it won't help with actual pagination, because there is no LastEvalutedKey.
